pwndbg/storage.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Provides persistent volatile and non-volatile storage of the results of
+a function call.  Volatile storage uses the temporary directory, and is
+generally cleared each reboot, or as the OS dictates.
+
+Cached results are stored per-architecture, i.e. arm and i386 caches
+are independent.  Cached results are also stored per-argument.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import functools
+import inspect
+import os
+import pickle
+import shelve
+import tempfile
+import time
+
+import pwndbg.arch
+import pwndbg.config
+
+nonvolatile = pwndbg.config.Parameter('pwndbg-cache-nonvolatile', '~/.pwndbg', 'Non-volatile cache directory for Pwndbg storage')
+volatile = pwndbg.config.Parameter('pwndbg-cache-volatile', tempfile.gettempdir(), 'Volatile cache directory for Pwndbg storage')
+
+def name(function):
+    mod_name = function.__module__
+    func_name = function.__name__
+    return mod_name + '.' + func_name
+
+def path(function, directory=volatile):
+    return os.path.join(str(directory), name(function))
+
+class Cached(object):
+    """Function decorator that persistently caches the results of the function it decorates."""
+    function = None
+    type = volatile
+    shelf = None
+
+    def __init__(self, function):
+        self.function = function
+        functools.update_wrapper(self, function)
+
+    def __setup_shelf(self):
+        """Deferred initialization allows customization of the paths by the user"""
```

Comment:
No, it's specifically so that the `pwndbg.config.Parameter` (which defines the filesystem path) can be changed by the user before we decide where to store the database